File: chapter_2_linked_lists/python/2.3-delete_mid_node.py
SinglyLinkedList = __import__('singly_linked_list').SinglyLinkedList


class SinglyLinkedList_2_3(SinglyLinkedList):

    def del_middle(self, mid_node):
        """
        Deletes a node in the middle of a linked list
        O(1) time complexity
        """
        next_node = mid_node.next
        mid_node.value = next_node.value
        mid_node.next = next_node.next
        del(next_node)


sll = SinglyLinkedList_2_3()
sll.add_to_front(1)
sll.add_to_front(2)
node = sll.add_to_front(4)
sll.add_to_front(8)
sll.add_to_front(16)

sll.print_linked_list()  # 16-> 8-> 4-> 2-> 1
sll.del_middle(node)
sll.print_linked_list()  # 16-> 8-> 2-> 1

File: chapter_2_linked_lists/js/2.3-deleteMidNode.js
const { SinglyLinkedList } = require('./singlyLinkedList')

class SinglyLinkedList_2_3 extends SinglyLinkedList {
  /* Deletes a node in the middle of a linked list
     O(n) time complexity */
  deleteMiddle(node) {
    const nextNode = node.next
    node.value = nextNode.value
    node.next = nextNode.next
  }
}

const sll = new SinglyLinkedList_2_3()
sll.addNode(1)
sll.addNode(2)
sll.addNode(4)
const node = sll.addNode(8)
sll.addNode(16)
sll.addNode(32)
sll.addNode(64)
sll.printList() // 64 -> 32 -> 16 -> 8 -> 4 -> 2 -> 1
sll.deleteMiddle(node)
sll.printList() // 64 -> 32 -> 16 -> 4 -> 2 -> 1
